### PR TITLE
Allow embedded type parameters in references

### LINKED ISSUE #2655: Implement constructor tear-offs in dartdoc
Depends on dart-lang/sdk#46020.

We can add to dartdoc's internal reference scheme for unnamed constructor tearoffs based on the specification, here:  https://github.com/dart-lang/language/blob/master/accepted/future-releases/constructor-tearoffs/feature-specification.md.   Specifically, allowing comment references to use `[C.new]` for unnamed constructor tearoffs will simplify some things in the new reference lookup code.

[Edit by eernstg on Jul 8th: Updated link to feature spec.]

### PR DESCRIPTION
**Pull request by jcollins-g**
Continues work for #2655.

Given that type parameters have more interesting uses with tearoffs now, add the capacity to embed (and ignore) type parameters inside comment references.  They are ignored because a single reference can still only point to a single page.

### CODE REVIEW

File: test/end2end/model_special_cases_test.dart [hunk 1]
```diff
@@ -203,6 +203,31 @@ void main() {
         expect(referenceLookup(constructorTearoffs, 'Ft.new'),
             equals(MatchingLinkResult(Fnew)));
       });
+
+      test('we can use (ignored) type parameters in references', () {
+        expect(referenceLookup(E, 'D<String>.new'),
```
Comment (srawlins) [inline test/end2end/model_special_cases_test.dart:208]: A type argument can also be a prefixed identifier, like `<core.String>`.

By all means, does not have to be addressed in this CL, but maybe a TODO.

**Comment by srawlins** [comment]
Everything in this PR refers to "type _parameters_" but is really concerned with "type _arguments_", at least as far as the analyzer API and language spec are concerned. In that verbiage:

* Type parameter is an element specifying a generic type on a class, function type, alias, mixin, extension, or function, (like `T` in `class A<T> {}`)
* Type argument is a type which is being used to instantiate a type parameter, e.g. `int` in `List<int> a` or `T` in `List<T> a`.
* Type variable is a classification of a static type, separating it from "class" or "interface type." E.g. `T` in T get first => ...`. I'm less certain of this one; it might just be an informal term I've seen used.

If "type parameters" are the term used throughout dartdoc though, than I think its fine to keep using in this PR for consistency. WDYT?

**Comment by jcollins-g** [comment]
@srawlins Part of the problem here is we allow people to pretend they are using arguments or type parameters interchangeably in dartdoc comment references and unfortunately that has spilled out not only into the rest of the code but into my brain when I talk about them.  I really should go back and fix all of that, I'll give that a try for this PR.

File: test/end2end/model_special_cases_test.dart (hunk 1 above)
Comment (jcollins-g) [reply test/end2end/model_special_cases_test.dart:208]: Yes.  Didn't add a test for that, but should be supported by the code here as well.

**Comment by jcollins-g** [comment]
renamed some things here.  PTAL

**Review by srawlins — APPROVED** [review]

**Pull request merged**